unblob/extractors/unromfs.py

```diff
@@ -0,0 +1,398 @@
+            os.mknod(inode.path, inode.mode, inode.dev)
+        elif inode.type == FS_TYPE.FIFO:
+            os.mkfifo(output_path, inode.mode)
+
+    def dump_fs(self):
+        # first we create files and directories
+        fd_inodes = {
+            k: v
+            for k, v in self.inodes.items()
+            if v.type in [FS_TYPE.FILE, FS_TYPE.DIRECTORY, FS_TYPE.FIFO, FS_TYPE.SOCKET]
+        }
+        for inode in sorted(fd_inodes.values(), key=lambda inode: inode.path):
+            self.create_inode(self.extract_root, inode)
+
+        # then we create links
+        links_inodes = {
+            k: v
+            for k, v in self.inodes.items()
+            if v.type in [FS_TYPE.SYMLINK, FS_TYPE.HARD_LINK]
+        }
+        for inode in sorted(links_inodes.values(), key=lambda inode: inode.path):
+            self.create_inode(self.extract_root, inode)
+
+        if os.geteuid() != 0:
+            print(
+                "root privileges are required to create block and char devices, skipping."
+            )
+            return
+
+        # then we create devices if we have enough privileges
+        dev_inodes = {
+            k: v
+            for k, v in self.inodes.items()
+            if v.type in [FS_TYPE.BLOCK_DEV, FS_TYPE.CHAR_DEV]
+        }
+        for inode in sorted(dev_inodes.values(), key=lambda inode: inode.path):
+            self.create_inode(self.extract_root, inode)
+
+    def __str__(self):
+        return f"signature: {self.signature}\nfull_size: {self.full_size}\nchecksum: {self.checksum}\nvolume_name: {self.volume_name}"
+
+
+@click.command(help="A tool to unpack RomFS filesystems.")
+@click.argument(
+    "romfs_file",
+    type=click.Path(path_type=Path, dir_okay=False, file_okay=True, exists=True),
+)
+@click.option(
+    "-e",
+    "--extract-dir",
+    "extract_root",
+    type=click.Path(path_type=Path, dir_okay=True, file_okay=False, resolve_path=True),
+    default=Path.cwd(),
+    help="Extract the files to this directory. Will be created if it doesn't exist.",
+)
+@click.option("-f", "--force", is_flag=True, help="Force overwrite.")
+@click.option("-v", "--verbose", is_flag=True, help="Verbose mode, enable debug logs.")
+def cli(romfs_file: Path, extract_root: Path, force: bool, verbose: bool):
+    configure_logger(verbose, extract_root)
+    if not romfs_file:
+        logger.error("No file provided")
+        return
+    logger.info("Start processing file", path=romfs_file)
+
+    if os.path.isdir(extract_root):
+        if force:
+            shutil.rmtree(extract_root)
+        else:
+            logger.warning(
+                f"extraction root {extract_root} already exists. Use -f operator to force overwrite."
+            )
+            return
+
+    with romfs_file.open("rb") as f:
+        header = RomFSHeader(f, extract_root, force)
+        header.validate()
+        header.recursive_walk(header.header_end_offset, None)
+        header.dump_fs()
+
+
+def main():
+    try:
```

Comment:
These shenanigans are not needed here. The only reason we have this in unblob is click's special handling of EOFError https://click.palletsprojects.com/en/8.0.x/exceptions/

Reply:
Great ! I'll get rid of it.